Add a run() method to run/wait for actors completion

daniel-zullo-frequenz opened the pull request:
Adds the `run()` method which awaits the completion of all actors. The method takes a variable number of actors as parameter and uses `asyncio.gather()` to wait for all actors to complete their tasks.

This will allow us to find better methods to wait for all actors to finish. It also simplify the way the actors are manually `join()` on the client side.

Partially addresses #36

daniel-zullo-frequenz committed:
Add run() method to simplify awaiting of actors completion

The manual join() of actors was causing issues with linting
tools such as pylint and mypy due to the indirection created
by the actor decorator. The new run() function simplifies and
improves the synchronization of actors on the client side.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz commented:
This is just a very early draft useful to discuss the changes and show the progress as there are other issues related to #36

daniel-zullo-frequenz committed:
Replace asyncio gather() with wait()

This will be only useful in the future if
the actor decorator does no longer manage
the lifecycle of the actor tasks.

Logging the tasks done in this case might contain
useful debugging information as they include
cancellation information.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz committed:
Fix logging exception in actor decorator

The exception information is added by default when calling
logger.exception and it is redundant to re-add it in the
message.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz committed:
Reset release notes

Keep the changes added after v0.18.0

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

christianparpart commented on src/frequenz/sdk/actor/_decorator.py:
```diff
@@ -207,10 +207,8 @@ async def _start_actor(self) -> None:
-                        "Actor (%s) crashed with error: %s", cls.__name__, err
-                    )
+                except Exception:  # pylint: disable=broad-except
+                    logger.exception("Actor (%s) crashed", cls.__name__)
```
Also applicable to more below. Why don't we use f-strings instead? This should be generally less error prone.

shsms replied:
it is because pylint prefers this,  and the logging package recommends this approach as the most optimal,  to have the least performance cost especially when log lines get discarded based on how the app is configured,  like debug logs are disabled in production, etc:  https://docs.python.org/3/howto/logging.html#optimization

leandro-lucarella-frequenz replied:
For logging it is recommended to use the built-in (old school) formatting as an *optimization*, as the interpolation will be only done if the message has to be emitted this way. If you use f-strings, interpolation will be done even if you don't end up printing the message. It is also unfortunate that uses old string interpolation with `%s` instead of the new `{}` interpolation, because you have to know both formats. If this optimization is really worth it or not, I don´t know, AFAIK nobody took the time to figure it out.

`pylint` has a rule for this and will complain if you use f-strings.

leandro-lucarella-frequenz replied:
Doing a quick search, it looks like the performance gain is really negligible, so I would be fine with disabling that `pylint` check globally and use f-strings instead.

* https://github.com/PyCQA/pylint/issues/2395

leandro-lucarella-frequenz replied:
I created a discussion about it so we avoid polluting this PR, for this PR lets leave the old style until we have a conclusion for the discussion.

* https://github.com/frequenz-floss/frequenz-sdk-python/discussions/206

leandro-lucarella-frequenz replied:
Can we mark this as resolved @christianparpart (I prefer to let people that initiated the conversation to mark it resolved, unless they said already that it was fine for them and there is someone else still having questions)

leandro-lucarella-frequenz commented on examples/power_distribution.py:
```diff
@@ -23,6 +23,7 @@
 from frequenz.sdk._data_ingestion import MicrogridData
 from frequenz.sdk._data_ingestion.formula_calculator import FormulaCalculator
 from frequenz.sdk.actor import actor
+from frequenz.sdk.actor import run as actor_run
```
Why are you aliasing this function? If you think users most likely will like to alias it, then I think we should change the function's name. Otherwise using this instead might be clear enough:

```py
from frequenz import actor

actor.run(...)
```

This is actually google's recommended way to import stuff, and I think it makes sense, but sadly we are not doing this consistently in the SDK, so I guess using a more explicit function name is OK too.

In any case, I would call this `run_actors` instead of `actor_run`.

daniel-zullo-frequenz replied:
I think I've added an alias because the module `examples/power_distribution` has a method `run()` but it is true that your suggestion would work. I'll work out the changes ~and I kept the import of the `actor` decorator as it is handy to use `@actor`~

daniel-zullo-frequenz replied:
I've removed the alias and now it uses `actor.run()` as suggested but the actor decorator is now called like this `@actor.actor`

leandro-lucarella-frequenz commented on src/frequenz/sdk/actor/_runner.py:
```diff
@@ -0,0 +1,41 @@
+        assert isinstance(actor, BaseActor), f"{actor} is not an instance of BaseActor"
+        actor_tasks.append(asyncio.create_task(actor.join()))
+
+    done, pending = await asyncio.wait(actor_tasks, return_when=asyncio.ALL_COMPLETED)
```
Doesn't `ALL_COMPLETED` guarantee that all task will be `done`? If so, I guess logging `pending` doesn't make a lot of sense, but since it is a debug log I'm fine with keeping it too. But if it is not guaranteed that all will be done, you need to put this in a loop that runs until all actors really stopped.

daniel-zullo-frequenz replied:
`ALL_COMPLETED` guarantees all tasks are finished or are cancelled. Indeed, `pending` does not make sense in this case. I was initially using `pending` when `return_when` was set to `FIRST_EXCEPTION`.

The actor decorator handles the actor restarts on exceptions (or actor termination) within a loop so I don't think we need to re-do it here at the moment.

I'll remove the pending variable and related logging. Though `asyncio.wait()` returns a tuple but I guess I can use `done, _ = await asyncio.wait(...)` to avoid a linting error of a non-used variable.

daniel-zullo-frequenz replied:
Removed pending, but I still kept the debugging messages for `done`

leandro-lucarella-frequenz replied:
Using `FIRST_COMPLETED` and looping might also be a good idea, so you get more *real-time* logging when actors finish running, and not just all the logging at the end. But we can leave this for a follow-up PR.

daniel-zullo-frequenz replied:
it makes sense to me to add the loop and use `FIRST_COMPLETED`, though I'd keep the inline comment that `currently the actor decorator manages the life-cycle of the actor tasks` to avoid misunderstandings (I guess this might change on follow-up PRs).

leandro-lucarella-frequenz reviewed (commented):
LGTM, any reason to keep this as a draft? Maybe add some test?

daniel-zullo-frequenz commented:
> Maybe add some test?

Yes, I'll add a couple of tests and remove the draft once it is done

daniel-zullo-frequenz committed:
Add unit tests for actor run utils

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz commented on src/frequenz/sdk/actor/_runner.py:
```diff
@@ -0,0 +1,38 @@
+            actor_tasks, return_when=asyncio.FIRST_COMPLETED
+        )
+        log.debug("Actor task done: %s", done)
+        log.debug("Actor task pending: %s", actor_tasks)
```
@leandro-lucarella-frequenz I'm happy to remove the logging if you think is not needed.
And what about `actor_tasks`? Should I better rename it to something like `actor_tasks_pending` instead? Pls let me know your thoughts

leandro-lucarella-frequenz replied:
I would just use `done, pending` as it is what we normally use. In any case, I would make done and pending symmetric in terms of name (`done`/`pendding`, `done_tasks`/`pending_tasks`, etc.)

daniel-zullo-frequenz commented:
Added a couple of simple unit tests and update the `wait()` function to use `FIRST_COMPLETED` within a loop instead.

mathias-baumann-frequenz commented on src/frequenz/sdk/actor/_runner.py:
```diff
@@ -0,0 +1,27 @@
+# License: MIT
+# Copyright © 2023 Frequenz Energy-as-a-Service GmbH
+
+"""Utility functions to run and synchronize the execution of actors."""
```
Hmm maybe a name like `_run_utils` or `async_utils` would be more descriptive?

mathias-baumann-frequenz replied:
(for the file)

leandro-lucarella-frequenz replied:
I would call it just `_run.py` if it only has `run` for now, to make it extra clear, but I'm fine with how it is now or `_run_utils`, `async_utils` seems to be too broad for what we have now and might be harder to expect the `run()` function to be found there.

daniel-zullo-frequenz replied:
Then I'll rename it to `_run_utils` as that's the name you kind of both suggested

daniel-zullo-frequenz replied:
done

leandro-lucarella-frequenz commented on tests/actor/test_runner.py:
```diff
@@ -0,0 +1,71 @@
+    sleepy_actors = [SleepyActor(f"sleepy_actor_{i}") for i in range(3)]
+    await run(*sleepy_actors)
+    for sleepy_actor in sleepy_actors:
+        assert sleepy_actor.is_joined, f"{sleepy_actor.name} was not joined"
```
You probably want to add a timeout here, otherwise if `run()` is misbehaving and waiting forever, tests will also wait forever. Same for the other test.

You should probably also check that the time elapsed in `run()` is more than `0.1` to make sure it also waited. Also, when you start playing with time, I strongly recommend using async-solipsism (makes the time pass in async code) and timemachine (fake the tiem), so tests don't really take that much time and we avoid flaky tests. 

Examples:

* https://github.com/frequenz-floss/frequenz-sdk-python/blob/v0.x.x/tests/timeseries/test_resampling.py
* https://github.com/frequenz-floss/frequenz-sdk-python/blob/v0.x.x/tests/actor/test_resampling.py

https://docs.python.org/3/library/asyncio-task.html#timeouts (warning: many of the timeout functions are only available in Python 3.11+).

daniel-zullo-frequenz replied:
ok, I'll amend the tests following the suggestions and examples

daniel-zullo-frequenz replied:
Updated based on the suggestion and examples. Please have another look at the test to check if that's pretty much what you have in mind

ela-kotulska-frequenz commented on src/frequenz/sdk/actor/_runner.py:
```diff
@@ -0,0 +1,27 @@
+from ._decorator import BaseActor
+
+
+async def run(*actors: Any) -> None:
```
Did you try to use `BaseActor` instead of `Any`?
async def run(*actors: `BaseActor`) -> None:

daniel-zullo-frequenz replied:
yes, I did try it and I'm afraid it didn't work. This is due to the way the actor decorator instantiate an internal class which is the `BaseActor`. This is why I've done the type check at runtime and I added an inline comment to highlight the reason of the type check at runtime

daniel-zullo-frequenz commented:
Updated based on suggestions from Luca and Mathias.

daniel-zullo-frequenz commented:
pylint was not happy with a unit test and I needed to update it

leandro-lucarella-frequenz commented on tests/actor/test_run_utils.py:
```diff
@@ -0,0 +1,112 @@
+import async_solipsism
+import pytest
+import time_machine
+from async_solipsism.socket import asyncio
```
I guess this was a wrong completion by "smart completion"? Shouldn't this be `import asyncio`?

daniel-zullo-frequenz replied:
Right, this should be `import asyncio`. As this was a copy/pasta from other test I will also amend the test where this import was copied from in a separate commit

daniel-zullo-frequenz replied:
done and fixed the import in another test in a separate commit

leandro-lucarella-frequenz commented on tests/actor/test_run_utils.py:
```diff
@@ -0,0 +1,112 @@
+from frequenz.sdk.actor import actor, run
+
+
+@pytest.fixture(autouse=True)
+def fake_loop() -> Iterator[async_solipsism.EventLoop]:
+    """Replace the loop with one that doesn't interact with the outside world."""
+    loop = async_solipsism.EventLoop()
+    yield loop
+    loop.close()
+
+
+@pytest.fixture
+def fake_time() -> Iterator[time_machine.Coordinates]:
+    """Replace real time with a time machine that doesn't automatically tick."""
+    with time_machine.travel(0, tick=False) as traveller:
+        yield traveller
```
Not necessarily for this PR, but we probably should move this to a tests/util module.

daniel-zullo-frequenz replied:
yes, we should. I'll create an issue and post it here once it's done

daniel-zullo-frequenz replied:
https://github.com/frequenz-floss/frequenz-sdk-python/issues/223

leandro-lucarella-frequenz commented on tests/actor/test_run_utils.py:
```diff
@@ -0,0 +1,112 @@
+        end_time = time.time()
+        self.virtual_time_taken_to_run = end_time - start_time
+        # pylint: disable=protected-access,no-member
+        await self._stop()  # type: ignore[attr-defined]
```
Mmm, calling `_stop()` inside `run()` looks fishy. Can't you call it outside of it?

daniel-zullo-frequenz replied:
yes, you're right. This shouldn't be like this. Either I'll call it from outside or will fix the actor decorator so that an actor is not restarted once it is finished without raising any exception

daniel-zullo-frequenz replied:
I've added a separate commit to fix the actor decorator, but the changes could be a bit controversial so I'll be also happy to create new issue/PR to discuss about it, and i can call _stop() in the test itself rather than in `run()`

leandro-lucarella-frequenz commented on tests/actor/test_run_utils.py:
```diff
@@ -0,0 +1,112 @@
+async def test_all_actors_done(fake_time: time_machine.Coordinates) -> None:
+    """Test the completion of all actors."""
+
+    async def sleep(duration: float) -> None:
+        fake_time.shift(duration)
+
+    with patch("asyncio.sleep", new=sleep):
```
Mmm, interesting, but I think this is not the correct approach.

Usually when faking time, you should "drive" the time manually and assert stuff while you make time pass. In this case, since you want to test something (`run`) waiting for something else (the actor run), you probably need a different task to monitor and make the time pass.

So the `await asyncio.sleep(self.duration_sec)` in the actor `run()` won't be executed until you make a `fake_time.shift()` of **at least** `duration_sec`. So you should assert for that to make sure `run()` is really waiting for the actor to finish.

daniel-zullo-frequenz replied:
Long story short this is my understanding, the `asyncio.sleep` was taken the real time rather than a negligible time given that the test was supposed to use a fake loop. Renaming the function from `fake_loop` to `event_loop` does the trick so that now the time taken is sleep is negligible and does not longer need to be patched.

Moreover we agreed in a call how to better handle the time is god mode :) and verify it. So I still need to amend the tests accordingly.

Also I'll need to patch other tests where `fake_loop` function name was used, will add these changes in a separate commit

leandro-lucarella-frequenz commented on src/frequenz/sdk/actor/_run_utils.py:
```diff
@@ -0,0 +1,38 @@
+    pending_tasks = []
+    # Check that each actor is an instance of BaseActor at runtime,
+    # due to the indirection created by the actor decorator.
+    for actor in actors:
+        assert isinstance(actor, BaseActor), f"{actor} is not an instance of BaseActor"
+        pending_tasks.append(asyncio.create_task(actor.join()))
+
+    # Currently the actor decorator manages the life-cycle of the actor tasks
+    while pending_tasks:
+        done_tasks, pending_tasks = await asyncio.wait(
+            pending_tasks, return_when=asyncio.FIRST_COMPLETED
+        )
+        log.debug("Actor task done: %s", done_tasks)
```
I think you should check if the task ended with an exception and log an error in that case (same if it was cancelled).

```suggestion
    for actor in actors:
        assert isinstance(actor, BaseActor), f"{actor} is not an instance of BaseActor"
        pending_tasks.append(asyncio.create_task(actor.join())),
             name=str(actor)

    # Currently the actor decorator manages the life-cycle of the actor tasks
    while pending_tasks:
        done_tasks, pending_tasks = await asyncio.wait(
            pending_tasks, return_when=asyncio.FIRST_COMPLETED
        )
        # This should always be only one task, but we handle many for extra safety
        for task in done:
            # Cancellation needs to be checked first, otherwise the other methods could raise a CancelledError
            if task.cancelled():
	         log.info("The actor %s was cancelled", task.get_name())
            elif exception := task.exception():
	         log.error("The actor %s was finished due to an uncaught exception", task.get_name(), exc_info=exception)
	     else:
	         log.info("The actor %s finished normally", task.get_name())
```

daniel-zullo-frequenz replied:
This is a great idea. We might want to set a name for the actor at some point given that in the this case we are assigning names like `<frequenz.sdk.actor._decorator.actor.<locals>.ActorClass object at 0x7f1648012d10>`, just to keep it in mind...
I'll add the suggested changes

daniel-zullo-frequenz committed:
Fix the asyncio import in test

The module was imported from the wrong package.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz committed:
Rename function fake_loop to event_loop in tests

Using the function name `fake_loop` does not
replace the loop with one that doesn't interact
with the outside world.
So the function needs to be named `event_loop`
instead.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz committed:
Fix infinite loop in actor decorator for finished actors

Previously, the actor decorator was causing an infinite loop
for actors that had finished normally (without exceptions)
even when the restart_limit attribute was set to zero.
This was particularly problematic in test cases where the
restart_limit attribute was intended to limit the number of
times an actor can be restarted.

To address this issue, the patch modified the code to consider
the restart_limit attribute in cases where exceptions different
than CancelledError are raised or when an actor finishes normally.
This ensures that the actor decorator does not cause an infinite
loop for finished actors, and the restart_limit attribute works
as expected.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz commented:
Updated based on the latest suggestions from Luca and also I fixed problems I found in tests in separate commits.

daniel-zullo-frequenz commented on RELEASE_NOTES.md:
```diff
@@ -2,17 +2,17 @@
-* A `MovingWindow` class has been added that consumes a data stream from a logical meter and updates an `OrderedRingBuffer`.
-* Add EVChargerPool implementation. It has only streaming state changes for ev chargers, now.
-* Add 3-phase current formulas: `3-phase grid_current` and `3-phase ev_charger_current` to the LogicalMeter.
 * A new class `SerializableRingbuffer` is now available, extending the `OrderedRingBuffer` class with the ability to load & dump the data to disk.
```
@ela-kotulska-frequenz is this correct to remove all the features above and keep `SerializableRingbuffer` as it hasn't been released yet?

ela-kotulska-frequenz replied:
No. They should stay. You should just add your features. :0

ela-kotulska-frequenz replied:
Yes, it is fine

ela-kotulska-frequenz commented on src/frequenz/sdk/actor/_decorator.py:
```diff
@@ -198,17 +198,23 @@ async def _start_actor(self) -> None:
                     logger.exception("Actor (%s) crashed", cls.__name__)
+                finally:
+                    if cancelled_error:
+                        raise cancelled_error
```
I don't understand how moving `raise cancelled_error` from `except asyncio.CancelledError as ex:` to `finally:` helped. :D Could you please explain it ?

ela-kotulska-frequenz replied:
Discussed. Thanks a lot.

ela-kotulska-frequenz reviewed (dismissed):
Approve from me. Thanks a lot!
Luca should look at it, too :)

leandro-lucarella-frequenz commented on src/frequenz/sdk/actor/_decorator.py:
```diff
@@ -198,19 +198,23 @@ async def _start_actor(self) -> None:
                     await super().run()
-                except asyncio.CancelledError:
+                except asyncio.CancelledError as ex:
+                    cancelled_error = ex
                     logger.debug("Cancelling actor: %s", cls.__name__)
-                    raise
-                except Exception as err:  # pylint: disable=broad-except
-                    logger.exception(
-                        "Actor (%s) crashed with error: %s", cls.__name__, err
-                    )
+                except Exception:  # pylint: disable=broad-except
+                    logger.exception("Actor (%s) crashed", cls.__name__)
+                finally:
+                    if cancelled_error:
+                        raise cancelled_error
```
This probably needs some re-work. I don't think you need to save the `cancelled_error`, you can re-raise it in the except clause as before.

I didn't have an in depth look to propose any alternatives, so I'd be also fine with leaving it as it is and improve it later if it's too complicated/time consuming.

daniel-zullo-frequenz replied:
Yes, that's right. The exception should be raised in the except clause. However I'd need a flag to check for CancelledError in the finally clause as the break disables the exception given the current logic in the exception handling. Actually I'm leaning towards to change the current retry-mechanism so that there is not logic needed when handling the exception

leandro-lucarella-frequenz commented on src/frequenz/sdk/actor/_run_utils.py:
```diff
@@ -0,0 +1,51 @@
+
+from ._decorator import BaseActor
+
+log = logging.getLogger(__name__)
```
```suggestion
_logger = logging.getLogger(__name__)
```

daniel-zullo-frequenz replied:
will update, thanks!

leandro-lucarella-frequenz commented on tests/actor/test_resampling.py:
```diff
@@ -26,7 +26,7 @@
 
 
 @pytest.fixture(autouse=True)
```
You can remove the `autouse=True` too, as this magically replaces the event loop in all tests. Maybe add a comment instead so people doesn't get confused and think that adding or removing `autouse` have any effect?

daniel-zullo-frequenz replied:
Right, I'll add a comment then

leandro-lucarella-frequenz commented on tests/timeseries/test_resampling.py:
```diff
@@ -38,7 +38,7 @@
 
 
 @pytest.fixture(autouse=True)
```
Same here.

leandro-lucarella-frequenz reviewed (commented):
Mostly minor comments, except for the one about exception handling in the actor's `run()` method.

daniel-zullo-frequenz committed:
Unset autouse for solipsism loop in tests

Setting autouse has no effect for solipsism event_loop() as
the method replaces the event loop for all tests in the file
where it is used.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz commented:
Updated based on latest suggestions from Luca.

daniel-zullo-frequenz commented:
Fixed a couple of issues related to mypy but the tests still failing :(
I need to check what's going on

daniel-zullo-frequenz commented:
It was a problem with pylint, it is fixed now.

daniel-zullo-frequenz commented:
Rebase onto latest v0.x.x after the flaky tests have been fixed

ela-kotulska-frequenz commented on src/frequenz/sdk/actor/_run_utils.py:
```diff
@@ -19,9 +22,30 @@ async def run(*actors: Any) -> None:
         assert isinstance(actor, BaseActor), f"{actor} is not an instance of BaseActor"
-
-    await asyncio.gather(*(actor.join() for actor in actors))
+        pending_tasks.add(asyncio.create_task(actor.join(), name=str(actor)))  # type: ignore
```
You should stop created tasks before raising assert error.

daniel-zullo-frequenz replied:
The life-cycle of the actor is managed by the actor decorator in the current state so it might not be a good idea to stop the tasks here.
Though we can create 2 separated loops `for actor in actors`, the first one will check for a programming error (actor is a BaseActor) which raises and exits the function without creating any tasks and the second loops will create tasks only after validation is passed. We can get rid of the first loop once the actor decorator is converted to a class. What do you think?
Just in case , this is what i mean:

```python
    for actor in actors:
        assert isinstance(actor, BaseActor), f"{actor} is not an instance of BaseActor"
        
    for actor in actors:
        pending_tasks.add(asyncio.create_task(actor.join(), name=str(actor)))  # type: ignore
```

leandro-lucarella-frequenz replied:
This looks more resilient to me :+1:

daniel-zullo-frequenz replied:
Then I'll update it like that

ela-kotulska-frequenz replied:
Looks good:)

leandro-lucarella-frequenz commented on src/frequenz/sdk/actor/_decorator.py:
```diff
@@ -201,25 +201,23 @@ async def _start_actor(self) -> None:
+                self.restart_limit is None or number_of_restarts <= self.restart_limit
+            ):
                 try:
+                    if number_of_restarts > 0:
+                        logger.info("Restarting actor: %s", cls.__name__)
```
I would move this outside of the `try:` as you should only catch errors produced by the actor here, not your own code.

Other than that, great! This looks much cleaner now! :+1:

daniel-zullo-frequenz replied:
Makes sense. Updated

leandro-lucarella-frequenz reviewed (commented):
Almost there :)

daniel-zullo-frequenz commented:
Updated based on latest suggestions from Ela and Luca

daniel-zullo-frequenz commented:
It seems the branch v0.x.x was recently updated and I'll need to rebase it

daniel-zullo-frequenz commented:
Rebased onto latest v0.x.x

daniel-zullo-frequenz commented:
@ela-kotulska-frequenz I'm afrait the PR lost your approval after rebasing it onto the lates v0.x.x recently updated. Would you mind to re-approve it if LGTY?
